src/task/future.rs

```diff
@@ -0,0 +1,388 @@
+// Copyright 2019 TiKV Project Authors. Licensed under Apache-2.0.
+
+//! A [`Future`].
+
+use crate::{LocalSpawn, RemoteSpawn};
+
+use std::cell::UnsafeCell;
+use std::future::Future;
+use std::marker::PhantomData;
+use std::mem::ManuallyDrop;
+use std::pin::Pin;
+use std::sync::atomic::{AtomicU8, Ordering::SeqCst};
+use std::sync::Arc;
+use std::task::{Context, Poll, RawWaker, RawWakerVTable, Waker};
+use std::{fmt, mem};
+
+/// A [`Future`] task.
+pub struct Task<Remote, Extras> {
+    status: AtomicU8,
+    future: UnsafeCell<Pin<Box<dyn Future<Output = ()> + Send + 'static>>>,
+    remote: Remote,
+    extras: UnsafeCell<Extras>,
+}
+
+/// A [`Future`] task cell.
+pub struct TaskCell<Remote, Extras>(Arc<Task<Remote, Extras>>);
+
+// Safety: It is ensured that `future` and `extras` are always accessed by
+// only one thread at the same time.
+unsafe impl<Remote: Sync, Extras> Sync for Task<Remote, Extras> {}
```

Comment:
I change the original code to `unsafe impl` for `Task`.
PTAL the safety insurance @BusyJay @nrc

Reply:
I think `Extras` needs to be `Send` to make `TaskCell::mut_extras` safe, otherwise it will allow to send `Extras` to other thread no matter `Extras` is `Send` or not.

Reply:
I don't get it. `TaskCell` is `Send` and `Sync` only if `Extras` is `Send`, so `extras` won't be sent to other threads.

Reply:
Never mind, seems `TaskCell::mut_extras` takes a mut reference, it should be safe.